File: config_1_uniform.py
import sys
import csv
import os
# Check if the correct number of arguments is provided
if len(sys.argv) != 3:
    print("Usage: python3.10 config_1_uniform.py N factor (the bandwidth per GPU is factor*960GBps)")
    sys.exit(1)  # Exit with an error code
# Extract command-line arguments
N = int(sys.argv[1])
factor = int(sys.argv[2]) # the bandwidth per GPU is factor*960GBps

from ogdf_python import *
# Tile_length = 100.0
cppinclude("ogdf/fileformats/GraphIO.h")
cppinclude("ogdf/orthogonal/OrthoLayout.h")
cppinclude("ogdf/planarity/PlanarSubgraphFast.h")
cppinclude("ogdf/planarity/PlanarizationLayout.h")
cppinclude("ogdf/planarity/SubgraphPlanarizer.h")
cppinclude("ogdf/planarity/VariableEmbeddingInserter.h")

G = ogdf.Graph()
# Create the nodes
nodes = [G.newNode() for _ in range(N)]
# Create two edges between each pair of nodes
num_edges=0
for i in range(N):
    for j in range(i+1, N):
        for f in range(factor):
            G.newEdge(nodes[i], nodes[j])
            num_edges+=1
            G.newEdge(nodes[j], nodes[i])
            num_edges+=1

GA = ogdf.GraphAttributes(G, ogdf.GraphAttributes.all)
# define graph attributes
GA.directed=False
for n in G.nodes:
    GA.label[n] = "%s" % n.index()
for e in G.edges:
    GA.arrowType[e] =0

# Initialize layout components
pl = ogdf.PlanarizationLayout()
crossMin = ogdf.SubgraphPlanarizer()
ps = ogdf.PlanarSubgraphFast[int]()
ps.runs(100)
ves = ogdf.VariableEmbeddingInserter()
ves.removeReinsert(ogdf.RemoveReinsertType.All)

# Set up PlanarizationLayout components
crossMin.setSubgraph(ps)
crossMin.setInserter(ves)
pl.setCrossMin(crossMin)

ol = ogdf.OrthoLayout()
# ol.separation(10.0)
# ol.cOverhang(10.0)
pl.setPlanarLayouter(ol)
pl.call(GA)

ogdf.GraphIO.drawSVG(GA, f"N_{N}_{factor}_TBps_config_1_uniform.svg")  # for SVG format

num_crossings=pl.numberOfCrossings()
ave_crossing_per_wg=num_crossings/num_edges

# [N, Theta, ave_crossing_per_wg] in the csv file
data = [N, factor*960, ave_crossing_per_wg]
# Specify the file name
csv_file_name = "config_1_uniform_planarization.csv"
if os.path.exists(csv_file_name):
    # Writing to CSV file
    with open(csv_file_name, mode='a', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(data)
else:
    with open(csv_file_name, mode='w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(["N", "Theta[TBps]", "ave_crossing_per_wg"])
        writer.writerow(data)


# del G
# del GA
# del pl
# del ol
# del ves
# del crossMin

File: c++_native/config_1_uniform.cpp
#include <ogdf/basic/Graph.h>
#include <ogdf/layered/SugiyamaLayout.h>
#include <ogdf/fileformats/GraphIO.h>
#include <ogdf/orthogonal/OrthoLayout.h>
#include <ogdf/planarity/PlanarSubgraphFast.h>
#include <ogdf/planarity/PlanarizationLayout.h>
#include <ogdf/planarity/SubgraphPlanarizer.h>
#include <ogdf/planarity/VariableEmbeddingInserter.h>
#include <filesystem>

using namespace ogdf;

int main(int argc, char *argv[]) {
    // Check if the correct number of arguments is provided
    if (argc != 3) {
        std::cerr << "Usage: ./executable N factor (the bandwidth per GPU is factor*(N-1)*4GBps*16(wavelengths)GBps)" << std::endl;
        return 1; // Exit with an error code
    }

    // Extract command-line arguments
    int N = std::stoi(argv[1]);
    int factor = std::stoi(argv[2]); // the bandwidth per GPU is factor*(N-1)*4GBps*16(wavelengths)GBps

    // Create the graph
    Graph G;

    // Create the nodes
    NodeArray<node> nodes(G);
    for (int i = 0; i < N; ++i) {
        nodes[i] = G.newNode();
    }

    // Create edges
    int num_edges = 0;
    for (int i = 0; i < N; ++i) {
        for (int j = i + 1; j < N; ++j) {
            for (int f = 0; f < factor; ++f) {
                G.newEdge(nodes[i], nodes[j]);
                ++num_edges;
                G.newEdge(nodes[j], nodes[i]);
                ++num_edges;
            }
        }
    }

    // Set up graph attributes
    GraphAttributes GA(G, GraphAttributes::all);
    // GA.directed = false;
    for (node n : G.nodes) {
        GA.label(n) = std::to_string(n->index());
    }
    for (edge e : G.edges) {
        GA.arrowType(e) = EdgeArrow::None;
    }

    // Initialize layout components
    PlanarizationLayout pl;
    SubgraphPlanarizer *crossMin = new SubgraphPlanarizer;
    PlanarSubgraphFast<int> *ps = new PlanarSubgraphFast<int>;
    ps->runs(100);
	VariableEmbeddingInserter *ves = new VariableEmbeddingInserter;
	ves->removeReinsert(RemoveReinsertType::All);

    // Set up PlanarizationLayout components
    crossMin->setSubgraph(ps);
    crossMin->setInserter(ves);
    pl.setCrossMin(crossMin);

    OrthoLayout *ol = new OrthoLayout;
    // ol->separation(20.0);
	// ol->cOverhang(0.4);
    pl.setPlanarLayouter(ol);
    pl.call(GA);

    // Draw the graph to SVG file
    GraphIO::drawSVG(GA, "svgs/N_" + std::to_string(N) + "_" + std::to_string(factor) +
                                   "_TBps_config_1_uniform.svg");

    int num_crossings = pl.numberOfCrossings();
    int ave_crossing_per_wg = round((num_crossings)/num_edges);

    // [N, Theta, ave_crossing_per_wg] in the csv file
    std::vector<int> data = {N, factor * (N-1)*4*16, ave_crossing_per_wg};

    // Specify the file name
    const std::string csv_file_name = "config_1_uniform_planarization.csv";

    std::ofstream file;
    if (std::filesystem::exists(csv_file_name)) {
        // Writing to CSV file
        file.open(csv_file_name, std::ios::app);
        file << data[0] << "," << data[1] << "," << data[2] << "\n";
    } else {
        file.open(csv_file_name);
        file << "N,Theta[GBps],ave_crossing_per_wg\n";
        file << data[0] << "," << data[1] << "," << data[2] << "\n";
    }
    file.close();

    G.clear();
    // nodes.clear()
    // GA.clearAllBends
    return 0;
}